src/hotspot/cpu/x86/c2_CodeStubs_x86.cpp

```diff
@@ -73,6 +73,83 @@ void C2EntryBarrierStub::emit(C2_MacroAssembler& masm) {
   __ jmp(continuation(), false /* maybe_short */);
 }
 
+int C2FastUnlockLightweightStub::max_size() const {
+  return 128;
+}
+
+void C2FastUnlockLightweightStub::emit(C2_MacroAssembler& masm) {
+  assert(_t == rax, "must be");
+
+  Label restore_held_monitor_count_and_slow_path;
+
+  { // Restore lock-stack and handle the unlock in runtime.
+
+    __ bind(_push_and_slow_path);
+#ifdef ASSERT
+    // The obj was only cleared in debug.
+    __ movl(_t, Address(_thread, JavaThread::lock_stack_top_offset()));
+    __ movptr(Address(_thread, _t), _obj);
+#endif
+    __ addl(Address(_thread, JavaThread::lock_stack_top_offset()), oopSize);
+  }
+
+  { // Restore held monitor and slow path.
+
+    __ bind(restore_held_monitor_count_and_slow_path);
+    // Restore held monitor count.
+    __ increment(Address(_thread, JavaThread::held_monitor_count_offset()));
+    // increment will always result in ZF = 0 (no overflows).
+    // continuation is the slow_path.
+    __ jmp(continuation());
+  }
+
+  { // Handle monitor medium path.
+
+    __ bind(_check_successor);
+
+    Label fix_zf_and_unlocked;
+    const Register monitor = _mark;
+
+#ifndef _LP64
+    // The owner may be anonymous, see comment in x86_64 section.
+    __ movptr(Address(monitor, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner)), _thread);
+    __ jmpb(restore_held_monitor_count_and_slow_path);
+#else // _LP64
+    // The owner may be anonymous and we removed the last obj entry in
+    // the lock-stack. This loses the information about the owner.
+    // Write the thread to the owner field so the runtime knows the owner.
```

Comment:
I'm confused by this comment.  We get here if the monitor is inflated, so we didn't remove it from the lock stack.

Reply:
True. This comment was written when there was an explicit monitor check before the CAS that jumped to inflated. I am not sure if there is a situation where the owner is anonymous here now. 

It should be invariant that if a thread's lock stack does not contain the oop, performs an unlock/monitorexit, the monitor is inflated and the owner is not anonymous. 

At all places in the runtime when removing the oops from the lock stack the owner field is fixed.

And in the emitted code the oop is pushed back to the lock stack incase of a failed unlock. 

There may be worth keeping this, and in the slow path after the CAS failed, check if it failed because of inflation, fix the owner field and jump back to the inflated fast path without transitioning to VM.

Reply:
I'm still confused by this.   I added this, because iiuc this is an invariant here?

```
+#ifdef ASSERT
+    Label skip;
+    __ testb(Address(monitor, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner)), (int32_t) ObjectMonitor::ANONYMOUS_OWNER);
+    __ jccb(Assembler::equal, skip);
+    __ stop("owner is anonymous");
+    __ bind(skip);
+#endif
+
```

I don't suggest adding this but am still trying to understand this comment.

Reply:
I ran through tier1-tier7 with the even stronger invariant that the owner is the thread. 
The current comment (in the code) is outdated. 

I plan to push a fixed version, just wanted to figure out what makes the most sense. 
Given that the monitor check is elided now, the fixing the owner field should be removed. 
Alternatively let the slow path check for monitor after the CAS failed, and jump back to the
inflated case. In this case the comment and the fixing of the owner field would be important. 

It is not immediately obvious to me that the alternative is worth it. Because we removed hashcode causing inflation the only point where this code would avoid going into the runtime by checking for inflated after a CAS fail is if another thread is in entering on the object monitor in parallel and we can do a direct handoff. Otherwise it would be on the entry / cxq queue and going into the runtime is required. 
The pragmatic solution would be to remove the store and add the assert for this PR. And create an RFE to evaluate this. 

Ran through with this code.
```c++
#ifdef ASSERT
    Label correct_owner;
    __ movptr(_t, Address(monitor, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner)));
    __ cmpptr(_t, _thread);
    __ jccb(Assembler::equal, correct_owner);
    __ testptr(_t, Address(_t)); // Crash on bad address
    __ stop("Bad owner");
    __ bind (correct_owner);
#endif
```

I will rebase and push the changes tomorrow.

Reply:
I like how with your patch all the knowledge about the anonymous owner is in synchronizer.cpp, which is why this comment and code would be nicer to be absent.

Reply:
> Is the lock definitely never inflated due to hash code in this case now?

It never inflates a locked object. The hash code race that may re-inflate a monitor is with deflation. Where it tries to install a hash code on an inflated monitor, but the monitor gets deflated between the two reads of the markWord, the thread that tries to install the hash code will create a new ObjectMonitor. [JDK-8323724](https://bugs.openjdk.org/browse/JDK-8323724) RFE is created for that issue.

Reply:
Hmm now that I think about it can still happen that it inflates it. If a third thread enters at the same time.

Reply:
But regardless all paths which commit to a pop also fixes the owner. So we do not need the anonymous owner fix in fast_unlock.

Reply:
I removed the comment and the anonymous owner fix. Did not add an ASSERT check as the runtime asserts the same thing.